### PR TITLE
NativeAOT-LLVM: cast unsigned ints to floats using uitofp

### PR DESCRIPTION
**Pull request by yowl**
This PR fixes a bug when casting ints to floating points.  It assumed signed.  If the source is an unsigned int, then make the cast unsigned (https://llvm.org/docs/LangRef.html#uitofp-to-instruction) .

Adds a test using a byte > 0x7f

cc @SingleAccretion

### CODE REVIEW

**Comment by jkotas** [comment]
@SingleAccretion Could you please take a look at this change? I guess it might have slipped through your inbox.

File: src/coreclr/tools/aot/ILCompiler.LLVM/CodeGen/ILToLLVMImporter.cs [hunk 1]
```diff
@@ -4035,7 +4037,15 @@ private void ImportConvert(WellKnownType wellKnownType, bool checkOverflow, bool
             }
             else
             {
-                LLVMValueRef converted = CastIfNecessary(loadedValue, GetLLVMTypeForTypeDesc(destType), value.Name(), wellKnownType == WellKnownType.UInt64 /* unsigned is always false, so check for the type explicitly */);
+                TypeDesc sourceType = value.Type;
```
Comment (SingleAccretion) [inline src/coreclr/tools/aot/ILCompiler.LLVM/CodeGen/ILToLLVMImporter.cs]: Who calls `ImportConvert`?

Edit: I see, it is called from the shared logic.

File: src/coreclr/tools/aot/ILCompiler.LLVM/CodeGen/ILToLLVMImporter.cs [hunk 2]
```diff
@@ -4035,7 +4037,15 @@ private void ImportConvert(WellKnownType wellKnownType, bool checkOverflow, bool
...
+
+                LLVMValueRef converted = CastIfNecessary(loadedValue, GetLLVMTypeForTypeDesc(destType), value.Name(),
+                    wellKnownType == WellKnownType.UInt64 /* unsigned is always false, so check for the type explicitly */
+                    || sourceType.IsWellKnownType(WellKnownType.Byte)
+                    || sourceType.IsWellKnownType(WellKnownType.UInt16)
+                    || sourceType.IsWellKnownType(WellKnownType.UInt32)
+                    || sourceType.IsWellKnownType(WellKnownType.UInt64)
+                    );
```
Comment (SingleAccretion) [inline src/coreclr/tools/aot/ILCompiler.LLVM/CodeGen/ILToLLVMImporter.cs]: How does the ILImporter stack work -- does it keep the precise types?

If it does, it does not look correct to look at them. E. g.
```
ldobj int64
conv.r.un
```
Should still be treated as an unsigned cast.

> `/* unsigned is always false, so check for the type explicitly */`

Is that true? I see it should be `true` for `conv.r.un`.

Comment (yowl) [reply src/coreclr/tools/aot/ILCompiler.LLVM/CodeGen/ILToLLVMImporter.cs]: Hmm, no that's not true, let me try to clean this up.

Comment (yowl) [reply src/coreclr/tools/aot/ILCompiler.LLVM/CodeGen/ILToLLVMImporter.cs]: There were 2 reported problems, using a byte as an index and casting a byte to a float (for an alpha channel).  But looking at the code I don't see why I can't just pass the right value from the op code.

File: src/coreclr/tools/Common/TypeSystem/IL/ILImporter.cs [hunk 3]
```diff
@@ -581,7 +581,7 @@ private void ImportBasicBlock(BasicBlock basicBlock)
                         ImportConvert(WellKnownType.Double, false, false);
                         break;
                     case ILOpcode.conv_u4:
-                        ImportConvert(WellKnownType.UInt32, false, false);
+                        ImportConvert(WellKnownType.UInt32, false, true);
```
Comment (jkotas) [inline src/coreclr/tools/Common/TypeSystem/IL/ILImporter.cs:584]: Could you please submit this fix to dotnet/runtime as well?

File: src/coreclr/tools/aot/ILCompiler.LLVM/CodeGen/ILToLLVMImporter.cs (hunk 2 above)
Comment (SingleAccretion) [inline src/coreclr/tools/aot/ILCompiler.LLVM/CodeGen/ILToLLVMImporter.cs]:
```suggestion
                LLVMValueRef converted = CastIfNecessary(loadedValue, GetLLVMTypeForTypeDesc(destType), value.Name(), unsigned);
```
Presumably that should now work?

I also see there is:
```
            else if ((toStoreKind == LLVMTypeKind.LLVMDoubleTypeKind || toStoreKind == LLVMTypeKind.LLVMFloatTypeKind) &&
                valueTypeKind == LLVMTypeKind.LLVMIntegerTypeKind)
            {
                //TODO: keep track of the TypeDesc so we can call BuildFPToUI when the integer is unsigned
                typedToStore = builder.BuildFPToSI(source, valueType, "CastFloatSI" + (name ?? ""));
            }
```
In `CastIfNecessary` that, could, presumably also be fixed.

**Comment by yowl** [comment]
I've put code to widen the bytes/shorts to ints before the `conv`.  Added another test.  Hopefully this fits the IL semantics better.

Example LLVM is now
```ll
  store i8 -1, i8* %alpha_local0_, align 1, !dbg !389570
  %Loadloc0_ = load i8, i8* %alpha_local0_, align 1, !dbg !389571
  %CastZIntloc0_ = zext i8 %Loadloc0_ to i32, !dbg !389571
  %CastSIToFloatloc0_ = sitofp i32 %CastZIntloc0_ to float, !dbg !389571
  %fpextop2 = fpext float %CastSIToFloatloc0_ to double, !dbg !38957
```

**Comment by yowl** [comment]
ugg, I've messed up the branches somehow... moving to WIP

File: src/coreclr/tools/aot/ILCompiler.LLVM/CodeGen/ILToLLVMImporter.cs [hunk 4]
```diff
@@ -4026,27 +4027,34 @@ private void ImportConvert(WellKnownType wellKnownType, bool checkOverflow, bool
...
             //TODO checkOverflow - r_un & r_4, i & i_un
             StackEntry value = _stack.Pop();
+            TypeDesc stackType = value.Type;
+
             TypeDesc destType = GetWellKnownType(wellKnownType);
 
             // Load the value and then convert it instead of using ValueAsType to avoid loading the incorrect size
             LLVMValueRef loadedValue = value.ValueAsType(value.Type, _builder);
+            LLVMValueRef widenedStackValue = CastIfNecessary(loadedValue, GetLLVMTypeForTypeDesc(WidenBytesAndShorts(value.Type)), value.Name(),
+                stackType.IsWellKnownType(WellKnownType.Byte) ||
```
Comment (SingleAccretion) [inline src/coreclr/tools/aot/ILCompiler.LLVM/CodeGen/ILToLLVMImporter.cs:4038]: Missing `Bool`?

**Review by SingleAccretion — APPROVED** [review]
LGTM modulo one comment.

**Comment by jkotas** [comment]
@yowl @SingleAccretion Thank you!

**Pull request merged**